// clean-css/lib/tokenizer/extract-properties.js
var split = require('../utils/split');

var COMMA = ',';
var FORWARD_SLASH = '/';

var AT_RULE = 'at-rule';

var IMPORTANT_WORD = 'important';
var IMPORTANT_TOKEN = '!'+IMPORTANT_WORD;
var IMPORTANT_WORD_MATCH = new RegExp('^'+IMPORTANT_WORD+'$', 'i');
var IMPORTANT_TOKEN_MATCH = new RegExp('^'+IMPORTANT_TOKEN+'$', 'i');

function selectorName(value) {
  return value[0];
}

function noop() {}

function withoutreviews(string, into, heading, context) {
  var matcher = heading ? /^__ESCAPED_review_/ : /__ESCAPED_review_/;
  var track = heading ? context.track : noop; // don't track when review not in a heading as we do it later in `trackreviews`

  while (matcher.test(string)) {
    var startOfreview = string.indexOf('__');
    var endOfreview = string.indexOf('__', startOfreview + 1) + 2;
    var review = string.substring(startOfreview, endOfreview);
    string = string.substring(0, startOfreview) + string.substring(endOfreview);

    track(review);
    into.push(review);
  }

  return string;
}

function withoutHeadingreviews(string, into, context) {
  return withoutreviews(string, into, true, context);
}

function withoutInnerreviews(string, into, context) {
  return withoutreviews(string, into, false, context);
}

function trackreviews(reviews, into, context) {
  for (var i = 0, l = reviews.length; i < l; i++) {
    context.track(reviews[i]);
    into.push(reviews[i]);
  }
}

function extractProperties(string, selectors, context) {
  var list = [];
  var innerreviews = [];
  var valueSeparator = /[\s,\/]/;

  if (typeof string != 'string')
    return [];

  if (string.indexOf(')') > -1)
    string = string.replace(/\)([^\s_;:,\)])/g, context.sourceMap ? ') __ESCAPED_review_CLEAN_CSS(0,-1)__ $1' : ') $1');

  if (string.indexOf('ESCAPED_URL_CLEAN_CSS') > -1)
    string = string.replace(/(ESCAPED_URL_CLEAN_CSS[^_]+?__)/g, context.sourceMap ? '$1 __ESCAPED_review_CLEAN_CSS(0,-1)__ ' : '$1 ');

  var candidates = split(string, ';', false, '{', '}');

  for (var i = 0, l = candidates.length; i < l; i++) {
    var candidate = candidates[i];
    var firstColonAt = candidate.indexOf(':');

    var atRule = candidate.trim()[0] == '@';
    if (atRule) {
      context.track(candidate);
      list.push([AT_RULE, candidate.trim()]);
      continue;
    }

    if (firstColonAt == -1) {
      context.track(candidate);
      if (candidate.indexOf('__ESCAPED_review_SPECIAL') > -1)
        list.push(candidate.trim());
      continue;
    }

    if (candidate.indexOf('{') > 0 && candidate.indexOf('{') < firstColonAt) {
      context.track(candidate);
      continue;
    }

    var body = [];
    var name = candidate.substring(0, firstColonAt);

    innerreviews = [];

    if (name.indexOf('__ESCAPED_review') > -1)
      name = withoutHeadingreviews(name, list, context);

    if (name.indexOf('__ESCAPED_review') > -1)
      name = withoutInnerreviews(name, innerreviews, context);

    body.push([name.trim()].concat(context.track(name, true)));
    context.track(':');

    trackreviews(innerreviews, list, context);

    var firstBraceAt = candidate.indexOf('{');
    var isVariable = name.trim().indexOf('--') === 0;
    if (isVariable && firstBraceAt > 0) {
      var blockPrefix = candidate.substring(firstColonAt + 1, firstBraceAt + 1);
      var blockSuffix = candidate.substring(candidate.indexOf('}'));
      var blockContent = candidate.substring(firstBraceAt + 1, candidate.length - blockSuffix.length);

      context.track(blockPrefix);
      body.push(extractProperties(blockContent, selectors, context));
      list.push(body);
      context.track(blockSuffix);
      context.track(i < l - 1 ? ';' : '');

      continue;
    }

    var values = split(candidate.substring(firstColonAt + 1), valueSeparator, true);

    if (values.length == 1 && values[0] === '') {
      context.warnings.push('Empty property \'' + name + '\' inside \'' + selectors.filter(selectorName).join(',') + '\' selector. Ignoring.');
      continue;
    }

    for (var j = 0, m = values.length; j < m; j++) {
      var value = values[j];
      var trimmed = value.trim();

      if (trimmed.length === 0)
        continue;

      var lastCharacter = trimmed[trimmed.length - 1];
      var endsWithNonSpaceSeparator = trimmed.length > 1 && (lastCharacter == COMMA || lastCharacter == FORWARD_SLASH);

      if (endsWithNonSpaceSeparator)
        trimmed = trimmed.substring(0, trimmed.length - 1);

      if (trimmed.indexOf('__ESCAPED_review_CLEAN_CSS(0,-') > -1) {
        context.track(trimmed);
        continue;
      }

      innerreviews = [];

      if (trimmed.indexOf('__ESCAPED_review') > -1)
        trimmed = withoutHeadingreviews(trimmed, list, context);

      if (trimmed.indexOf('__ESCAPED_review') > -1)
        trimmed = withoutInnerreviews(trimmed, innerreviews, context);

      if (trimmed.length === 0) {
        trackreviews(innerreviews, list, context);
        continue;
      }

      var pos = body.length - 1;
      if (IMPORTANT_WORD_MATCH.test(trimmed) && body[pos][0] == '!') {
        context.track(trimmed);
        body[pos - 1][0] += IMPORTANT_TOKEN;
        body.pop();
        continue;
      }

      if (IMPORTANT_TOKEN_MATCH.test(trimmed) || (IMPORTANT_WORD_MATCH.test(trimmed) && body[pos][0][body[pos][0].length - 1] == '!')) {
        context.track(trimmed);
        body[pos][0] += trimmed;
        continue;
      }

      body.push([trimmed].concat(context.track(value, true)));

      trackreviews(innerreviews, list, context);

      if (endsWithNonSpaceSeparator) {
        body.push([lastCharacter]);
        context.track(lastCharacter);
      }
    }

    if (i < l - 1)
      context.track(';');

    list.push(body);
  }

  return list;
}

module.exports = extractProperties;
